### PR TITLE
[Core] Deprecating block manager v1 and make block manager v2 default

### PR DESCRIPTION
**Pull request by KuntaiDu**
This PR deprecates block manager v1 and makes block manager v2 the default to simplify the code path.

This is supported by this [benchmark](https://docs.google.com/document/d/1XxYUFai07ta5rE7OdtCVhLJ5J0oAxEqrGgarFdjv0Zc/edit?usp=sharing), where block manager v2 is <2% slower than block manager v1 on Llama 8B when no prefix hit, and has significant speedup upon full prefix hit.

Summary of changes:
- Leave `--use-v2-block-manager` in the EngineArgs for compatibility
- Remove `use_v2_block_manager` flag in all tests and configs (except during initialization), so that the value change of `use-v2-block-manager` has no effect on vLLM behavior.

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Adding or changing kernels</h3>
<p>Each custom kernel needs a schema and one or more implementations to be registered with PyTorch.</p>
<ul>
    <li>Make sure custom ops are registered following PyTorch guidelines: <a href="https://pytorch.org/tutorials/advanced/cpp_custom_ops.html#cpp-custom-ops-tutorial">Custom C++ and CUDA Operators</a> and <a href="https://docs.google.com/document/d/1_W62p8WJOQQUzPsJYa7s701JXt0qf2OfLub2sbkHOaU">The Custom Operators Manual</a></li>
    <li>Custom operations that return <code>Tensors</code> require meta-functions. Meta-functions should be implemented and registered in python so that dynamic dims can be handled automatically. See above documents for a description of meta-functions.</li>
    <li>Use <a href="https://pytorch.org/docs/stable/library.html#torch.library.opcheck"><code>torch.libary.opcheck()</code></a> to test the function registration and meta-function for any registered ops.  See <code>tests/kernels</code> for examples.</li>
    <li>When changing the C++ signature of an existing op, the schema must be updated to reflect the changes.</li>
    <li>If a new custom type is needed, see the following document: <a href="https://docs.google.com/document/d/18fBMPuOJ0fY5ZQ6YyrHUppw9FA332CpNtgB6SOIgyuA">Custom Class Support in PT2</a>.
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by comaniac** [comment]
FYI: @sroy745 has #8678 verifying the functional correctness. Could you folks coordinate on this?

**Comment by KuntaiDu** [comment]
Sure! This PR will be a draft PR until @sroy745 verifies all the tests. I will also talk to @sroy745 and see if I can help.

**Comment by comaniac** [comment]
> Sure! This PR will be a draft PR until @sroy745 verifies all the tests. I will also talk to @sroy745 and see if I can help.

Thanks! @sroy745 has identified some failed tests and is fixing them. We could have a tracking issue and work together on fixing them.

**Comment by sroy745** [comment]
> > Sure! This PR will be a draft PR until @sroy745 verifies all the tests. I will also talk to @sroy745 and see if I can help.
> 
> Thanks! @sroy745 has identified some failed tests and is fixing them. We could have a tracking issue and work together on fixing them.

I filed #8718 to track the unit test failures. I am currently looking at the test_scheduler.py failures.

**Comment by KuntaiDu** [comment]
Add ready to trigger full set of CI and see which test fails

**Comment by sroy745** [comment]
fyi I have one pr in flight https://github.com/vllm-project/vllm/pull/8824 which fixes the last of the know test failures that I found earlier.

**Comment by sroy745** [comment]
@KuntaiDu all my tests fixes for the known breakages are merged. Can you please re-run the tests for this pr again and verify. I tried doing it for my pr (https://github.com/vllm-project/vllm/pull/8678) several times but somehow the tests get stuck beyond the first few format checks.

**Comment by KuntaiDu** [comment]
> @KuntaiDu all my tests fixes for the known breakages are merged. Can you please re-run the tests for this pr again and verify. I tried doing it for my pr (#8678) several times but somehow the tests get stuck beyond the first few format checks.

ofc ofc!

**Comment by sroy745** [comment]
@KuntaiDu this pr https://github.com/vllm-project/vllm/pull/8678 making the BlockManagerV2 the default is ready. Can you PTAL.

**Comment by comaniac** [comment]
@KuntaiDu we should merge #8678 first.

**Comment by KuntaiDu** [comment]
> @KuntaiDu we should merge #8678 first.

I am not a committer of vLLM, let me ping Kaichao

**Comment by comaniac** [comment]
> > @KuntaiDu we should merge #8678 first.
> 
> 
> 
> I am not a committer of vLLM, let me ping Kaichao

Oh I didn't notice that...then can you approve the PR? And I can merge it.

**Comment by KuntaiDu** [comment]
> > > @KuntaiDu we should merge #8678 first.
> > 
> > 
> > I am not a committer of vLLM, let me ping Kaichao
> 
> Oh I didn't notice that...then can you approve the PR? And I can merge it.

Done 😄

File: vllm/engine/arg_utils.py
Comment (comaniac) [inline vllm/engine/arg_utils.py:377]: Another option we could consider is adding a deprecate warning without removing v1 in a separate PR. And we release that PR first, followed by this PR that completely removes v1 along with this flag. Would like to hear others' thoughts.

**Review by comaniac — COMMENTED** [review]
This PR looks pretty exciting lol
Let's gather feedback from others and finalize the deprecation strategy.

cc @simon-mo @zhuohan123 @WoosukKwon

File: vllm/engine/arg_utils.py
Comment (KuntaiDu) [reply vllm/engine/arg_utils.py:377]: Agree, let's see how people react towards making block manager v2 default, and then fully remove block manager v1 after people are OK with block manager v2 as default.

**Comment by comaniac** [comment]
From @simon-mo 

* When the flag is set to false (i.e. people intentionally set it to v1), we raise an overridable error (similar to how we handle beam search) to comment why do they still need v1, to see any incompatibilities (we've already seen one in encoder-decoder...)
* We can wait for a week or two after the release, if there’s no major issue/incompatibility, then delete the v1 code and the flag.

**Comment by sroy745** [comment]
@comaniac @KuntaiDu I will send a pr shortly with the change suggested in this comment https://github.com/vllm-project/vllm/pull/8704#issuecomment-2394786536

**Comment by sroy745** [comment]
Sent out pr https://github.com/vllm-project/vllm/pull/9149 to address comment https://github.com/vllm-project/vllm/pull/8704#issuecomment-2394786536

**Comment by zhuohan123** [comment]
@KuntaiDu When do you think we can merge this PR?

**Comment by comaniac** [comment]
We may need to merge this after the next release. v2 is now the default and we already raise error when v1 is specified.

**Comment by KuntaiDu** [comment]
Let me fix the comments and conflicts by the end of this weekend

File: vllm/core/block_manager.py
Comment (KuntaiDu) [reply vllm/core/block_manager.py:20]: I am planning to call it "SelfAttnBlockSpaceManager". Does this sound like a right name for you?

Comment (comaniac) [reply vllm/core/block_manager.py:20]: SGTM

**Comment by comaniac** [comment]
0.6.3 has been released so we can continue working on this PR to deprecate v1.

**Comment by KuntaiDu** [comment]
@comaniac I fixed merge conflicts and removed some unnecessary flags and functions for block manager v1. PTAL

File: .buildkite/test-pipeline.yaml
Comment (comaniac) [inline .buildkite/test-pipeline.yaml:91]: I'm not sure if this is split by purpose to support conditional testing. cc @khluu to confirm

File: tests/core/block/e2e/test_correctness.py
Comment (comaniac) [inline tests/core/block/e2e/test_correctness.py:50]: Then maybe just rename the test to be something like `test_block_manager_with_preemption`?

File: tests/core/block/e2e/test_correctness.py
Comment (comaniac) [inline tests/core/block/e2e/test_correctness.py:222]:
```suggestion
def test_chunked_prefill_block_manager(baseline_llm_generator,
```
Ditto to other similar cases.

File: tests/core/block/e2e/test_correctness.py
Comment (comaniac) [inline tests/core/block/e2e/test_correctness.py:304]: Ditto. We should rename the test to reflect what's actually being tested.

File: vllm/engine/arg_utils.py
Comment (comaniac) [inline vllm/engine/arg_utils.py:1018]:
```suggestion
                "behavior. Please remove --use-v2-block-manager in your engine argument. "
```

**Review by comaniac — APPROVED** [review]
Otherwise LGTM. Also cc @sroy745 for review.

File: tests/core/block/e2e/test_correctness.py
Comment (sroy745) [inline tests/core/block/e2e/test_correctness.py:46]: nit- Please consider updating the comment "Verify block manager v2 produces same outputs as block manager v1, even when there is preemption." to "Verify block manager produces same outputs even when there is preemption"

File: tests/core/block/e2e/test_correctness.py
Comment (sroy745) [inline tests/core/block/e2e/test_correctness.py:300]: nit - Please consider removing reference to block manager v1 in the comment of the test.

File: .buildkite/test-pipeline.yaml
Comment (KuntaiDu) [reply .buildkite/test-pipeline.yaml:91]: The PR that turns `pytest -v -s core` to multiple test commands seem to be in PR #9149 from @sroy745 so I think it is probably not for the purpose of supporting conditional testing but I am not sure.

File: vllm/engine/arg_utils.py
Comment (sroy745) [inline vllm/engine/arg_utils.py:379]: nit - Do you want to call out BlockSpaceManagerV2 is same as  SelfAttnBlockSpaceManager by saying something like "and block manager v2 (SelfAttnBlockSpaceManager) is now the default " since we have renamed the class.

**Review by sroy745 — COMMENTED** [review]
Thanks for the pr!! LGTM.

File: .buildkite/test-pipeline.yaml
Comment (sroy745) [reply .buildkite/test-pipeline.yaml:91]: I had added that to selectively add the env variable VLLM_ALLOW_DEPRECATED_BLOCK_MANAGER_V1 only to tests which needed it. I added it explicitly to make sure I am not missing any tests which continue to use BlockManagerV1 and pass because this env gets set. I think we can remove it.

File: tests/core/block/e2e/test_correctness.py
Comment (KuntaiDu) [reply tests/core/block/e2e/test_correctness.py]: Fixed

File: vllm/engine/arg_utils.py
Comment (KuntaiDu) [reply vllm/engine/arg_utils.py]: Fixed.

File: tests/core/block/e2e/test_correctness.py
Comment (KuntaiDu) [reply tests/core/block/e2e/test_correctness.py:45]: Changed! TY for suggestion.

File: tests/core/block/e2e/test_correctness.py
Comment (KuntaiDu) [reply tests/core/block/e2e/test_correctness.py:298]: Changed! TY for suggestion.

File: vllm/engine/arg_utils.py
Comment (KuntaiDu) [reply vllm/engine/arg_utils.py]: Good point --- just adjusted the docstring.

File: tests/core/block/e2e/test_correctness.py
Comment (KuntaiDu) [reply tests/core/block/e2e/test_correctness.py]: Oh yes ur right. Fixed.

**Pull request merged**
